crates/sui-graphql-rpc/src/types/transaction_block/tx_lookups.rs

```diff
@@ -0,0 +1,468 @@
+        {
+            return Ok(None);
+        }
+
+        Ok(Some(Self::new(
+            lo,
+            hi,
+            page.after().map(|x| x.tx_sequence_number),
+            page.before().map(|x| x.tx_sequence_number),
+            scan_limit,
+            page.is_from_front(),
+        )))
+    }
+
+    /// The lower bound `tx_sequence_number`` of the range to scan within. This defaults to the min
+    /// tx_sequence_number of the checkpoint bound. If a cursor is provided, the lower bound is
+    /// adjusted to the larger of the two. The resulting value is additionally modified by the
+    /// `scan_limit` if `is_from_front` is false.
+    pub(crate) fn scan_lo(&self) -> u64 {
+        let adjusted_lo = self.after.map_or(self.lo, |a| std::cmp::max(self.lo, a));
+
+        if self.is_from_front {
+            adjusted_lo
+        } else {
+            std::cmp::max(
+                adjusted_lo,
+                self.scan_hi()
+                    .saturating_sub(self.scan_limit.unwrap_or(self.hi)),
+            )
+        }
+    }
+
+    /// The upper bound `tx_sequence_number` of the range to scan within. This defaults to the max
+    /// tx_sequence_number of the checkpoint bound. If a cursor is provided, the upper bound is
+    /// adjusted to the smaller of the two. The resulting value is additionally modified by the
+    /// `scan_limit` if `is_from_front` is true.
+    pub(crate) fn scan_hi(&self) -> u64 {
+        let adjusted_hi = self.before.map_or(self.hi, |b| std::cmp::min(self.hi, b));
+
+        if self.is_from_front {
+            std::cmp::min(
+                adjusted_hi,
+                self.scan_lo()
+                    .saturating_add(self.scan_limit.unwrap_or(adjusted_hi)),
+            )
+        } else {
+            adjusted_hi
+        }
+    }
+
+    /// If the query result does not have a previous page, check whether the scan limit is within
+    /// the initial tx_sequence_number range.
+    pub(crate) fn scan_has_prev_page(&self) -> bool {
+        if self.after.unwrap_or(0) >= self.hi {
+            return false;
+        }
+
+        self.scan_lo() > self.lo
+    }
+
+    /// If the query result does not have a next page, check whether the scan limit is within the
+    /// initial tx_sequence_number range.
+    pub(crate) fn scan_has_next_page(&self) -> bool {
+        if self.before.unwrap_or(self.hi) <= self.lo {
+            return false;
+        }
+
+        self.scan_hi() < self.hi
+    }
+}
+
+impl TransactionBlockFilter {
+    /// Returns the count of `function`, `kind`, `recv_address`, `input_object`, and
+    /// `changed_object`.
+    pub(crate) fn complex_filters(&self) -> usize {
+        [
+            self.function.is_some(),
+            self.kind.is_some(),
+            self.recv_address.is_some(),
+            self.input_object.is_some(),
+            self.changed_object.is_some(),
+        ]
+        .iter()
+        .filter(|&is_set| *is_set)
```

Comment:
Same deal here about `into_iter` and avoiding the dereference.

```suggestion
        .into_iter()
        .filter(|is_set| is_set)
```